src/ChunkedStreamDecoder.php

```diff
@@ -0,0 +1,174 @@
+    protected $maxBufferLength;
+
+    /**
+     * @var bool
+     */
+    protected $nextChunkIsLength = true;
+
+    /**
+     * @var ReadableStreamInterface
+     */
+    protected $stream;
+
+    /**
+     * @param ReadableStreamInterface $stream
+     * @param int $maxBufferLength
+     */
+    public function __construct(ReadableStreamInterface $stream, $maxBufferLength = self::MAX_BUFFER_LENGTH)
+    {
+        $this->stream = $stream;
+        $this->stream->on('data', array($this, 'handleData'));
+        Util::forwardEvents($this->stream, $this, [
+            'error',
+        ]);
+        $this->maxBufferLength = $maxBufferLength;
+    }
+
+    /** @internal */
+    public function handleData($data)
+    {
+        $this->buffer .= $data;
+
+        do {
+            $bufferLength = strlen($this->buffer);
+            $continue = $this->iterateBuffer();
+            $iteratedBufferLength = strlen($this->buffer);
+        } while (
+            $continue &&
+            $bufferLength !== $iteratedBufferLength &&
+            $iteratedBufferLength > 0 &&
+            strpos($this->buffer, static::CRLF) !== false &&
+            $iteratedBufferLength <= $this->maxBufferLength
+        );
+
+        if ($iteratedBufferLength >= $this->maxBufferLength) {
+            $this->emit('error', [
+                new Exception('The current buffer is longer then the ' . $this->maxBufferLength / 1024 . 'KB we have set as maximum'),
+            ]);
+            $this->close();
+            return false;
+        }
+    }
+
+    protected function iterateBuffer()
+    {
+        if (strlen($this->buffer) <= 1) {
+            return false;
+        }
+
+        if ($this->nextChunkIsLength) {
+            $crlfPosition = strpos($this->buffer, static::CRLF);
+            if ($crlfPosition === false) {
+                return false; // Chunk header hasn't completely come in yet
+            }
+            $this->nextChunkIsLength = false;
+            $lengthChunk = substr($this->buffer, 0, $crlfPosition);
+            if (strpos($lengthChunk, ';') !== false) {
+                list($lengthChunk) = explode(';', $lengthChunk, 2);
+            }
+            if (dechex(hexdec($lengthChunk)) !== $lengthChunk) {
+                $this->emit('error', [
+                    new Exception('Unable to validate "' . $lengthChunk . '" as chunk length header"'),
+                ]);
+                $this->close();
+                return false;
+            }
+            $this->remainingLength = hexdec($lengthChunk);
+            if ($this->remainingLength >= $this->maxBufferLength) {
+                $this->emit('error', [
+                    new Exception('Expected chunk length is longer then the ' . $this->maxBufferLength / 1024 . 'KB we have set as maximum'),
+                ]);
+                $this->close();
+                return false;
+            }
+            $this->buffer = substr($this->buffer, $crlfPosition + 2);
+            return true;
```

Comment:
Yeah it should, but doesn't it make more sense to forward the event from the underlying stream rather then detecting it our self?